lit_llama/adapter.py

```diff
@@ -57,57 +93,81 @@ def forward(
+        # embedding size (C) dimension into num_heads (nh) and head_size (hs)
         head_size = C // self.n_head
         k = k.view(B, T, self.n_head, head_size)
         q = q.view(B, T, self.n_head, head_size)
         v = v.view(B, T, self.n_head, head_size)
 
-        q = apply_rope(q, rope)
-        k = apply_rope(k, rope)
+        # "Unlike standard positional embeddings rotary embeddings must be applied at every layer"
+        q = apply_rope(q, rope) # (B, T, nh, hs)
+        k = apply_rope(k, rope) # (B, T, nh, hs)
 
+        # now `key`, 'query` and `value` tensors are correctly represented: for each element in a batch (B)
+        # there is a number of heads (nh) and for each head there is a sequence of elements (T), each of them is
+        # represented by a vector of size `hs`
         k = k.transpose(1, 2)  # (B, nh, T, hs)
         q = q.transpose(1, 2)  # (B, nh, T, hs)
         v = v.transpose(1, 2)  # (B, nh, T, hs)
 
         if kv_cache is not None:
-            cache_k, cache_v = kv_cache
+            cache_k, cache_v = kv_cache # 2 * (B, nh, max_seq_length, hs)
             # check if reached token limit
             if input_pos[-1] >= max_seq_length:
+                # if we reached token limit and thus there is no space to put newly calculated `key` and `value`
+                # right next to cached ones, we need to rotate cache tensor along `max_seq_length` dimension by one
+                # element to the left: this will free up space for new `key` and `value`
                 input_pos = torch.tensor(max_seq_length - 1, device=input_pos.device)
                 # shift 1 position to the left
                 cache_k = torch.roll(cache_k, -1, dims=2)
                 cache_v = torch.roll(cache_v, -1, dims=2)
-            k = cache_k.index_copy(2, input_pos, k)
-            v = cache_v.index_copy(2, input_pos, v)
+            k = cache_k.index_copy(2, input_pos, k) # (B, nh, max_seq_length, hs)
+            v = cache_v.index_copy(2, input_pos, v) # (B, nh, max_seq_length, hs)
             kv_cache = k, v
 
         # efficient attention using Flash Attention CUDA kernels
-        y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0)
+        # ↓ (B, nh, T, hs) @ (B, nh, T, hs).mT --> (B, nh, T, T) @ (B, nh, T, hs) --> (B, nh, T, hs)
+        y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0) # (B, nh, T, hs)
 
+        # "Adapters are applied to the topmost layers to better tune the language
+        # representations with higher-level semantics".
         if self.block_idx >= self.adapter_start_layer:
             if adapter_kv_cache is not None:
-                ak, av = adapter_kv_cache
+                ak, av = adapter_kv_cache # 2 * (B, nh, aT, hs)
             else:
                 prefix = self.adapter_wte.weight.reshape(1, self.adapter_prompt_length, self.n_embd)
                 aT = prefix.size(1)
```

Comment:
We can replace `self.n_embd` calls with `C`

```suggestion
                aT = self.adapter_prompt_length
                prefix = self.adapter_wte.weight.reshape(1, aT, C)
```

Reply:
yeah, it's a neat fix. But, I'm not sure that it's a part of `adapter annotated` pull request.